File: src/nodejs/reducerCoordinator.js
/*

    Reducer Coordinator: State Manager and scheduler for BigLambda job 

    If all mappers are done:
        NReducers = Get the number of the reducers needed;
        If  NReducers == 1:
           Invoke single reducer and write the results;
           Job done;
        Else create event source for reducer;
    Else:
        Return
   
   
 Copyright 2014 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 SPDX-License-Identifier: MIT-0
 */


var async = require('async');
var AWS = require('aws-sdk');
var AWS = require('aws-sdk');
var async = require('async');
var config = require('config.json')('./jobinfo.json');
var es = require('event-stream');
var https = require('https');
var utils = require('./utils'); 
var zlib = require('zlib');
var s3 = new AWS.S3({ apiVersion: '2006-03-01' });
var utils = new utils();
var DEFAULT_REGION = "us-east-1";

/// STATES
var MAPPERS_DONE = 0;
var REDUCER_STEP = 1;

function invokeLambda(params, cb){
        var agent = new https.Agent();
        agent.maxSockets = 10;
        var lambda = new AWS.Lambda({
            region: DEFAULT_REGION,
            httpOptions: {
                agent: agent,
            timeout: 5*60000
            }
        });
        var status = lambda.invoke(params, function(err, obj){
            if (err){
                console.error(err);
                cb(err, null);
                return;
            }
            cb(null, obj);
        });
}

function getJobFiles(lparams, cb){
    s3.listObjects(lparams, function (err, data) {
        if(err){cb(err, null); return}
        cb(null, data.Contents);
    });
}


function writeReducerState(nReducers, nS3, bucket, fname, cb){
    var ts = new Date().getTime();
    var data = JSON.stringify({
                "reducerCount": nReducers, 
                "totalS3Files": nS3,
                "startTime": ts 
               });
    utils.writeToS3(s3, bucket, fname, data, {}, cb);
}

exports.handler = function(event, context, callback){
    
    var startTime = process.hrtime();
    const bucket = event.Records[0].s3.bucket.name;
    var key = decodeURIComponent(event.Records[0].s3.object.key.replace(/\+/g, ' '));
    console.log("Event Key", key);
    var jobId =  config.jobId; //"lmrtest2" ; //event.jobId;
    var mapCount = config.mapCount ; //267; //event.mapCount;
    var reducerFunctionName = config.reducerFunction; 
    var reducerHandler = config.reducerHandler; 

    //var mparams = {Bucket: bucket, Key: jobId + "/jobdata"};
    //s3.getObject(mparams, function(err, data){
    //    var contents = JSON.parse(data.Body.toString());
    //    console.log("----", contents);
    //});

    //// Reducer Coorindator Flow ////
    // 1. Check if all mappers are done
    // 2. Compute the # of the reducers needed
    // 3. If 1 reducer, then done
    // 4. Else create event source for reducer; Repeat!

    // Count mapper files
    function getMapperFiles(mappers){
        var ret = [];
        var re = /task\/mapper\/\d+/i;
        for (var i=0; i<mappers.length; i++){ 
            if (mappers[i].Key.match(re)){
                ret.push(mappers[i]);
            }
        }
        return ret;
    }

    function getReducerStateInfo(files, cb){
        var reducers = [];
        var maxIndex = 0;
        var reducerStep = false;
        var rIndex = 0;
   
        // CHECK if step is complete !

        // Check for the Reducer state
        for (var i=0; i<files.length; i++){
            var parts = files[i].Key.split('/');
            var re = /reducerstate.\d+/i;
            if(files[i].Key.match(re)){
                console.log("Found reducers -->", files[i].Key); 
                var idx = parseInt(files[i].Key.split('.')[1]);
                if (idx > rIndex)
                    rIndex = idx;
                reducerStep = true;
            }
        }
        
        // This is the first reducer
        if (reducerStep === false){
            var mapperOutputFiles = getMapperFiles(files);
            cb([MAPPERS_DONE, mapperOutputFiles]);
            return;
        }
        else{
            // Check if the reducer id done
            var params = {Bucket: bucket, Key: jobId + "/reducerstate." + rIndex}; 
            s3.getObject(params, function(err, data){
                var contents = JSON.parse(data.Body.toString());
                for (var i=0; i<files.length; i++){
                    // JobID/task/reducer{n}/1
                    var parts = files[i].Key.split('/');
                    if(parts.length < 3)
                        continue
                    var op = parts[2];
                    if (op.indexOf('reducer' + rIndex) >=0){
                        reducers.push(files[i]);
                    }
                }
                
                if(contents["reducerCount"] == reducers.length){
                    // Reducer step complete; move to the next reducer
                    cb([rIndex, reducers]);
                }else{
                    // not complete
                    cb([rIndex, []]);
                }

            });
            
        }
    }

    //
    function getReducerBatchSize(keys){
        // Try and create a balance distribution 
       
        //if (config.reducerBatchSize >= 2)
        //    return config.reducerBatchSize;
        
        //TODO: Read metadata & determine the # of reducers
        var batchSize = utils.computeBatchSize(keys, 1536); 
        if (batchSize >=2)
            return batchSize;
        return 2; // Atleast 2 so that reducer can terminate
    }

    function _check_job_done(files){
        var re = /result/i;
        for (var i=0; i<files.length; i++){ 
            if (files[i].Key.match(re)){
                return true; 
            }
        }
        return false;
    }
    
    function _get_keys(objs){
        keys = [];
        for(var i=0; i<objs.length; i++)
            keys.push(objs[i].Key);
        return keys;
    }

    function _get_s3_stats(cb){

        /// Get the average max & avg of 3 keys.
        params = {}
        s3.getObject(params, function(err, data){
            var size = data.ContentLength; 
            var linecount = data.Metadata.linecount; 
            cb(size, linecount);
        });
    }

    // Get Mapper Finished Count
    var lparams = {
        Bucket: bucket,
        Prefix: jobId 
    };

    getJobFiles(lparams, function(err, files){
        if(err){
            console.error("S3 Error: ", err);
            //?
        }

        // Check if Job is done
        if (true === _check_job_done(files)){
            // TODO: Unsubscribe from the event source
            callback(null, "Job done!!! Check the result file: " + jobId + "/result" );
            return;
        }

        // Stateless Coordinator logic
        var mapperKeys = getMapperFiles(files);
        console.log("Mappers Done so far ", mapperKeys.length);
        if (mapCount == mapperKeys.length){
            // All the mappers have finished, time to schedule the reducers

            getReducerStateInfo(files, function(stepInfo){

                var stepNumber = stepInfo[0];
                var reducerKeys = stepInfo[1];
               
                if(reducerKeys.length == 0){
                    callback(null, "Still waiting to finish Reducer step " + stepNumber);
                    return;
                }
                 
                // Compute this based on metadata of files
                var rBatchSize = getReducerBatchSize(reducerKeys); 
                
                console.log("Starting the the reducer step", stepNumber);
                console.log("Batch Size", rBatchSize);
                
                // Create Batch params for the Lambda function
                var rBatchParams = utils.batchCreator(reducerKeys, rBatchSize);
                
                /// Build the lambda parameters
                var nReducers = rBatchParams.length;
                var nS3 = nReducers * rBatchParams[0].length;
                var stepId = stepNumber +1;
                var lambdaBatchParams = [];
                for (var i=0; i<rBatchParams.length; i++){
                    var batch = rBatchParams[i];
                        var params = {
                            FunctionName: reducerFunctionName,
                            InvocationType: 'Event',
                            Payload: JSON.stringify({
                                bucket: bucket,
                                keys: _get_keys(batch),
                                jobBucket: bucket,
                                jobId: jobId,
                                nReducers: nReducers, 
                                stepId: stepId, 
                                reducerId: i 
                            })
                        };
                    lambdaBatchParams.push(params);
                }

                console.log("LBR", lambdaBatchParams);

                /// Stream the files from S3 & Reduce
                async.map(lambdaBatchParams, invokeLambda, function (err, rResults){
                    if (err){
                        console.error(err);
                        callback(null, 'Reducer invocation error');
                        return;
                    }

                    var fname = jobId + "/reducerstate." + stepId;
                    writeReducerState(nReducers, nS3, bucket, fname, function(err, data){
                        callback(null, 'Invoked Reducer Step');
                    });
                });


          });
        }else{
            ///
            console.log("Still waiting for all the mappers to finish ..");
            callback(null, "I am still waiting for all the mappers to finish ..");
        } 
    });
};

/*
 * var ev = {
    "bucket": "smallya-useast-1",
    "jobId": "jobid134",
    "mapCount": 1,
    "reducerFunctionName": "shell-exec",
    "reducerHandler": "index.handler"
}
var cx = {} 
cx.succeed = function(x){ console.log(x)}
callback = function(x, y){ console.log(x, y)}
exports.handler(ev, cx, callback);
*/

File: src/python/reducerCoordinator.py
'''
REDUCER Coordinator 

Copyright 2016 Amazon.com, Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: MIT-0
'''

import boto3
import json
import lambdautils
import random
import re
from io import StringIO
import time
import urllib

DEFAULT_REGION = "us-east-1";

### STATES
MAPPERS_DONE = 0;
REDUCER_STEP = 1;

### Helpers ###

# create an S3 session
s3 = boto3.resource('s3')
s3_client = boto3.client('s3')
lambda_client = boto3.client('lambda')

# Write to S3 Bucket
def write_to_s3(bucket, key, data, metadata):
    s3.Bucket(bucket).put_object(Key=key, Body=data, Metadata=metadata)

def write_reducer_state(n_reducers, n_s3, bucket, fname):
    ts = time.time()
    data = json.dumps({
                "reducerCount": '%s' % n_reducers, 
                "totalS3Files": '%s' % n_s3,
                "start_time": '%s' % ts 
               })
    write_to_s3(bucket, fname, data, {})

# Count mapper files
def get_mapper_files(files):
    ret = []
    for mf in files:
        if "task/mapper" in mf["Key"]:
            ret.append(mf)
    return ret

def get_reducer_batch_size(keys):
    #TODO: Paramertize memory size
    batch_size = lambdautils.compute_batch_size(keys, 1536, 1000)
    return max(batch_size, 2) # At least 2 in a batch - Condition for termination

def check_job_done(files):
    # TODO: USE re
    for f in files:
        if "result" in f["Key"]:
            return True
    return False

def get_reducer_state_info(files, job_id, job_bucket):
        
    reducers = [];
    max_index = 0;
    reducer_step = False;
    r_index = 0;
   
    # Check if step is complete
        
    # Check for the Reducer state
    # Determine the latest reducer step#
    for f in files:
        #parts = f['Key'].split('/');
        if "reducerstate." in f['Key']:
            idx = int(f['Key'].split('.')[1])
            if idx > r_index:
                r_index = idx
            reducer_step = True

    # Find with reducer state is complete 
    if reducer_step == False:
        # return mapper files
        return [MAPPERS_DONE, get_mapper_files(files)]
    else:
        # Check if the current step is done
        key = "%s/reducerstate.%s" % (job_id, r_index) 
        response = s3_client.get_object(Bucket=job_bucket, Key=key)
        contents = json.loads(response['Body'].read())
        
        # get reducer outputs
        for f in files:
            fname = f['Key']
            parts = fname.split('/')
            if len(parts) < 3:
                continue
            rFname = 'reducer/' + str(r_index)
            if rFname in fname:
                reducers.append(f)
        
        if int(contents["reducerCount"]) == len(reducers):
            return (r_index, reducers)
        else:
            return (r_index, [])

def lambda_handler(event, context):
    print("Received event: " + json.dumps(event, indent=2))

    start_time = time.time();

    # Job Bucket. We just got a notification from this bucket
    bucket = event['Records'][0]['s3']['bucket']['name']

    #key = urllib.unquote_plus(event['Records'][0]['s3']['object']['key'].encode('utf8'))
   
    config = json.loads(open('./jobinfo.json', "r").read())
    
    job_id =  config["jobId"]
    map_count = config["mapCount"] 
    r_function_name = config["reducerFunction"] 
    r_handler = config["reducerHandler"] 

    ### Get Mapper Finished Count ###
    
    # Get job files
    files = s3_client.list_objects(Bucket=bucket, Prefix=job_id)["Contents"]

    if check_job_done(files) == True:
        print("Job done!!! Check the result file")
        # TODO:  Delete reducer and coordinator lambdas
        return
    else:
        ### Stateless Coordinator logic
        mapper_keys = get_mapper_files(files)
        print("Mappers Done so far ", len(mapper_keys))

        if map_count == len(mapper_keys):
            
            # All the mappers have finished, time to schedule the reducers
            stepInfo = get_reducer_state_info(files, job_id, bucket)

            print("stepInfo", stepInfo)

            step_number = stepInfo[0];
            reducer_keys = stepInfo[1];
               
            if len(reducer_keys) == 0:
                print("Still waiting to finish Reducer step ", step_number)
                return
                 
            # Compute this based on metadata of files
            r_batch_size = get_reducer_batch_size(reducer_keys); 
                
            print("Starting the the reducer step", step_number)
            print("Batch Size", r_batch_size)
                
            # Create Batch params for the Lambda function
            r_batch_params = lambdautils.batch_creator(reducer_keys, r_batch_size);
                
            # Build the lambda parameters
            n_reducers = len(r_batch_params)
            n_s3 = n_reducers * len(r_batch_params[0])
            step_id = step_number +1;

            for i in range(len(r_batch_params)):
                batch = [b['Key'] for b in r_batch_params[i]]

                # invoke the reducers asynchronously
                resp = lambda_client.invoke( 
                        FunctionName = r_function_name,
                        InvocationType = 'Event',
                        Payload =  json.dumps({
                            "bucket": bucket,
                            "keys": batch,
                            "jobBucket": bucket,
                            "jobId": job_id,
                            "nReducers": n_reducers, 
                            "stepId": step_id, 
                            "reducerId": i 
                        })
                    )
                print(resp)

            # Now write the reducer state
            fname = "%s/reducerstate.%s"  % (job_id, step_id)
            write_reducer_state(n_reducers, n_s3, bucket, fname)
        else:
            print("Still waiting for all the mappers to finish ..")

'''
ev = {
    "Records": [{'s3': {'bucket': {'name': "smallya-useast-1"}}}],
    "bucket": "smallya-useast-1",
    "jobId": "jobid134",
    "mapCount": 1,
    "reducerFunctionName": "shell-exec",
    "reducerHandler": "index.handler"
}
lambda_handler(ev, {})
'''
